addons/sale/models/sale_order.py

```diff
@@ -553,10 +574,21 @@ def _onchange_commitment_date(self):
                 }
             }
 
-    @api.depends('pricelist_id')
-    def _compute_show_update_pricelist(self):
+    @api.depends('pricelist_id', 'currency_id')
+    def _compute_show_update_pricelist_or_currency(self):
         for order in self:
-            order.show_update_pricelist = order.order_line and order.pricelist_id and order._origin.pricelist_id != self.pricelist_id
+            order.show_update_pricelist_or_currency = order.order_line and (
+                order._origin.pricelist_id != self.pricelist_id
+                or order._origin.currency_id != self.currency_id
+            )
+
+    @api.onchange('currency_id')
+    def _onchange_currency_invalidate_pricelist(self):
+        for order in self:
+            if order.pricelist_id and order.currency_id != order.pricelist_id.currency_id:
+                new_currency = order.currency_id
+                order.pricelist_id = None  # this triggers the compute to change the currency
+                order.currency_id = new_currency
```

Comment:
Currency should be readonly when pricelist are enabled, and this onchange should not be necessary.

Reply:
Hm, that seems strange for me, because if you make a pricelist for euro, could you not still want to not use any pricelist for other currency? Or even in the same currency, if you make a pricelist for "special customers", you may still want to just not use it for every SO. That was why I made this change, so you could still choose to not use a pricelist, or use whatever currency you want without creating a pricelist for it.

Reply:
Well, I was thinking that when pricelists are enabled, we fall back on old behavior, but if you can do a smooth flow while supporting both, it should be fine :clap: